### ISSUE TITLE
Nonexistent issues get linked in commit messages and comments

### ISSUE THREAD

**Issue by Yakov5776** [go-gitea/gitea#30275]
### Description

Currently, when users mention nonexistent issues in commit messages or comments, the system generates a hyperlink for them. However, when hovered over, these links lead to a "Network error" tooltip due to a 404 error because the referenced issue doesn't exist. This behavior can be confusing and misleading for users.

**Potential Solutions:**
- **Sub-Optimal Solution:** Indicate that the issue does not exist when being hovered over in lieu of a Network error.
- **Optimal Solution:** Issues should only be highlighted if they exist within the repository like in GitHub.

What makes the second solution more difficult is because we can't be invoking SQL queries on every page render to check the existence of a referenced issue.
GitHub internally links the issue at the time of commit or comment, not at the page render and this can be confirmed by making a new initialized repository, referencing `#1` in a commit message **before** and **after** the first issue is created, and you'll only see the second commit (which is post-issue) get highlighted and linked.

### Gitea Version

v1.21.10

### Can you reproduce the bug on the Gitea demo site?

Yes

### Screenshots

![image](https://github.com/go-gitea/gitea/assets/28637006/8ea871f0-a635-4f33-9c30-30dfda7422b3)
![image](https://github.com/go-gitea/gitea/assets/28637006/1d51301d-41f6-49ab-85f8-d90e606f6813)

### How are you running Gitea?

Running a fresh instance via CLI from gh-codespaces.

### Database

SQLite

**Comment by KN4CK3R** [comment]
That's currently expected behaviour because the references are not checked because of performance reasons. Same for strings that look like commit hashes.

**Comment by silverwind** [comment]
What we could do is at time of creation of the commit or creation/edit of a comment containing issue refs, check if the referenced issue exists, and if it does not exist, don't perform the replacement in the comment or commit message.

Such a method is possible without any ill effects on performance and is also historically accurate so that one can not link to future issues.

**Comment by delvh** [comment]
Well… That does have a major drawback though, @silverwind:
If I understand your proposal correctly, it means we have to hardcode HTML into a comment.
I am not a fan of that at all, especially since it irreversibly changes what you entered.

**Comment by Yakov5776** [comment]
> Same for strings that look like commit hashes.

Nope, commit hashes are actually checked. I made sure to confirm this when making the issue.


https://github.com/go-gitea/gitea/assets/28637006/fe5056b7-d399-4aa7-a8b3-d23c0eab8d0d

**Comment by Yakov5776** [comment]
> Well… That does have a major drawback though, @silverwind: If I understand your proposal correctly, it means we have to hardcode HTML into a comment. I am not a fan of that at all, especially since it irreversibly changes what you entered.

Why HTML? it could just be hardcoding a _reference_ which gets translated into HTML on page render.

**Comment by silverwind** [comment]
> Well… That does have a major drawback though, @silverwind: If I understand your proposal correctly, it means we have to hardcode HTML into a comment. I am not a fan of that at all, especially since it irreversibly changes what you entered.

Indeed, that's a drawback. Also it would require commit messages to be stored in database which I think they are not currently.

**Comment by KN4CK3R** [comment]
And it will not highlight references which are not valid at commit time.

**Comment by silverwind** [comment]
On the other hand, existing instances with thousands of issues would not benefit from a fix that is done at commit/comment time, so likely doing it purely at render time is still ideal.

**Comment by Yakov5776** [comment]
Any updates? In the meantime, why don't we just handle "Not Found" responses?

**Comment by lunny** [comment]
Yes, maybe a quick fix is the frontend can handle 404 and return a better description to end users.

**Comment by silverwind** [comment]
We could at least detect the 404 and show "It looks like this issue does not exist", that'd at least get the ugly error away as a temporary solution.

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/31181 will hide the popup completely on non-existant issue, so after that, the only remaining task would be to not linkify these nonexistant issues in first place on backend.

**Comment by wxiaoguang** [comment]
For performance reason, backend render doesn't check every issue reference whether it exists (permission check is also needed, no permission also means "not found")

So the issue numbers will always get links. And we can fix the "network error" by showing "not found", proposed a new fix #35739
